lib/storage/gcs.js

```diff
@@ -0,0 +1,152 @@
+        endpoint = options.endpoint;
+        if (!endpoint.match(/^https?:/)) {
+          var defaultSecure = ((!options.port) || (options.port === 443));
+          var secure = options.secure || defaultSecure;
+          var port = options.port || 443;
+          var protocol = secure ? 'https://' : 'http://';
+          if (secure && (port === 443)) {
+            port = '';
+          } else if ((!secure) && (port === 80)) {
+            port = '';
+          } else {
+            port = ':' + port;
+          }
+          endpoint = protocol + endpoint + port;
+        }
+      }
+      // The storage client auth relies on the presence of the service account
+      // file path expressed in the environment variable
+      // GOOGLE_APPLICATION_CREDENTIALS and, of course, the presence of such file.
+      //
+      //
+      // See https://cloud.google.com/docs/authentication/getting-started
+      client = new Storage();
+      bucketName = options.bucket;
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      defaultTypes = require(__dirname + "/contentTypes.js");
+      if (options.contentTypes) {
+        _.extend(contentTypes, defaultTypes, options.contentTypes);
+      } else {
+        contentTypes = defaultTypes;
+      }
+      https = options.https;
+      cachingTime = options.cachingTime;
+      return callback(null);
+    },
+
+    copyIn: function(localPath, path, options, callback) {
+      async.waterfall([
+        function(callbackOne) {
+          var cleanPath = cleanKey(path);
+          var ext = extname(cleanPath);
+          if (ext.length) {
+            ext = ext.substr(1);
+          }
+          var contentType = contentTypes[ext];
+          if (!contentType) {
+            console.log(`unknown extension ${ext}, using mime octet-stream`);
+            contentType = 'application/octet-stream';
+          }
+
+          let cacheControl = 'no-cache';
+          if (cachingTime) {
+            cacheControl = 'public, max-age=' + cachingTime;
+          }
+          callbackOne(null, {
+            destination: cleanPath,
+            gzip: true,
+            metadata: {
+              cacheControl: cacheControl,
+              ContentType: contentType
+            }
+          });
+        },
+        function(uploadOptions, callbackTwo) {
+          client.bucket(bucketName).upload(localPath, uploadOptions).then(
+            function(apiResponse) {
+              // set the asset to public by default.
+              var partOne = apiResponse[0];
+              client.bucket(bucketName).file(partOne.metadata.name).makePublic().then(function (__) {
+                callbackTwo(null, apiResponse);
+              }, function (err) {
+                callback(err);
+              });
+            }).catch(function (reason) {
```

Comment:
This works, the promise chaining could be a little cleaner, this can be a flat chain of calls with just one catch() at the end and no use of a second argument to then() for readability. I can take care of that if you like, really appreciate this contribution.

Reply:
I can do that in the course of doing the other changes you requested, Thomas. :)